docs: how-to-purge-data (#473)
canonical/charmed-openstack-upgrader: docs/how-to/purge-data-on-shadow-table.rst

==========================================
Purge data on shadow tables
==========================================

> The purge steps is disable by default and plans to run before proceeding to actual upgrade steps.
> This can be turn on with the `--purge`` flag.

Reviewer: Some suggestions for readability.

```suggestion
The purge step is disabled by default.  When enabled, it will run before proceeding to actual upgrade steps.
This can be enabled with the ``--purge`` flag.
```

> This purge step is a performance optimisation, delete data from the shadow tables in nova database. The behavior is equal to run juju action `purge` on nova-cloud-controller unit, which can help to reduce the size of the db.

Reviewer: Note: rst uses double backticks for preformatted text - there are a couple of locations in this doc with single backticks.

> The ``purge-before-date`` flag is supported to delete the data older than the date provided. The date string format should be YYYY-MM-DD[HH:mm][:ss]. This flag is useful to retain recent data for debugging or data retention policies.

Reviewer: ```suggestion
The ``purge-before-date`` flag is supported to delete the data older than the date provided. The date string format should be ``YYYY-MM-DD[HH:mm][:ss]``. This flag is useful to retain recent data for debugging or data retention policies.
```

Usage examples
--------------

.. terminal::
   :input cou plan --purge --purge-before-date 2000-01-02

More information
----------------

- `nova-cloud-controller charm actions`_
- `nova-manage reference`_ - see `purge` subcommand
- :doc:`Archive old data <how-to/archive-old-data>`

> - :doc:`Nova data migration <explanation/nova-data-migration>`

Reviewer: Could you check the doc links please?  When I build the docs, sphinx emits some warnings:

```
/home/ubuntu/proj/canonical/charmed-openstack-upgrader/docs/explanation/nova-data-migration.rst:78: WARNING: unknown document: 'how-to/archive-old-data'
/home/ubuntu/proj/canonical/charmed-openstack-upgrader/docs/explanation/nova-data-migration.rst:79: WARNING: unknown document: 'how-to/purge-data-on-shadow-table'
/home/ubuntu/proj/canonical/charmed-openstack-upgrader/docs/how-to/archive-old-data.rst:57: WARNING: unknown document: 'how-to/purge-data-on-shadow-table'
/home/ubuntu/proj/canonical/charmed-openstack-upgrader/docs/how-to/archive-old-data.rst:58: WARNING: unknown document: 'explanation/nova-data-migration'
/home/ubuntu/proj/canonical/charmed-openstack-upgrader/docs/how-to/purge-data-on-shadow-table.rst:25: WARNING: unknown document: 'how-to/archive-old-data'
/home/ubuntu/proj/canonical/charmed-openstack-upgrader/docs/how-to/purge-data-on-shadow-table.rst:26: WARNING: unknown document: 'explanation/nova-data-migration'
```

.. LINKS
.. _nova-cloud-controller charm actions: https://charmhub.io/nova-cloud-controller/actions

> .. _nova-manage reference: https://docs.openstack.org/nova/rocky/cli/nova-manage.html

Reviewer: Important information that's missing in these docs is the benefits of archive + purge:

- performance optimisation for during the upgrade process (albeit small)
- reduce database size and disk usage
- improve performance of database queries during cloud operation because the db is smaller

I think this how-to is actually a good place for this information: you already mention performance optimisation, and I think it's important for a doc to include the 'why'.

Reviewer: I opened https://github.com/canonical/charmed-openstack-upgrader/issues/483 to track this.
